File: dotnet/src/Pagou.Examples.Core/Webhooks/WebhookProcessor.cs
using System.Text.Json.Serialization;

namespace Pagou.Examples.Core.Webhooks;

/// <summary>
/// The offloaded, slow half of webhook handling. It runs AFTER the fast 2xx ack.
/// Business state changes only on a confirmed event, and only after reconciling
/// against the API — the webhook body is a hint, the API is the source of truth.
/// </summary>
public sealed class WebhookProcessor
{
    private static readonly Dictionary<WebhookFamily, string> ResourcePath = new()
    {
        [WebhookFamily.Transaction] = "/v2/transactions",
        [WebhookFamily.Subscription] = "/v2/subscriptions",
        [WebhookFamily.Transfer] = "/v2/transfers",
    };

    private readonly PagouHttpClient _client;
    private readonly WebhookStore _store;

    public WebhookProcessor(PagouHttpClient client, WebhookStore store)
    {
        _client = client;
        _store = store;
    }

    public async Task ProcessEventAsync(WebhookEvent evt, CancellationToken ct = default)
    {
        if (!WebhookParser.IsConfirmedStateChange(evt.EventType))
        {
            Log.Info($"Ignoring non-confirming event {evt.EventType} ({evt.Id})");
            return;
        }
        if (string.IsNullOrEmpty(evt.ResourceId))
        {
            Log.Warn($"Confirmed event {evt.EventType} without a resource id — cannot reconcile.");
            return;
        }

        try
        {
            var result = await _client.RequestDataAsync<StatusOnly>(
                new RequestParams { Method = HttpMethod.Get, Path = $"{ResourcePath[evt.Family]}/{evt.ResourceId}" }, ct)
                .ConfigureAwait(false);
            _store.SetResourceState(evt.ResourceId, result.Data.Status);
            Log.Info($"Reconciled {evt.Family} {evt.ResourceId} → {result.Data.Status}");
        }
        catch (NotFoundException)
        {
            Log.Warn($"Resource {evt.ResourceId} not found during reconciliation.");
        }
        catch (Exception ex)
        {
            // Reconciliation failed after the ack: a real system would requeue this
            // event for a later retry rather than replaying side effects.
            Log.Error($"Reconciliation failed for {evt.Id}", new { message = ex.Message });
            throw;
        }
    }

    private sealed class StatusOnly
    {
        [JsonPropertyName("status")] public string Status { get; init; } = "";
    }
}

File: java/src/main/java/ai/pagou/examples/webhooks/WebhookProcessor.java
package ai.pagou.examples.webhooks;

import ai.pagou.examples.lib.Errors.NotFoundException;
import ai.pagou.examples.lib.Logger;
import ai.pagou.examples.lib.PagouHttpClient;
import ai.pagou.examples.lib.Request;
import ai.pagou.examples.webhooks.WebhookHandlers.Family;
import ai.pagou.examples.webhooks.WebhookHandlers.WebhookEvent;
import com.fasterxml.jackson.databind.JsonNode;
import java.util.Map;

// The offloaded, slow half of webhook handling. It runs AFTER the fast 2xx ack.
// Business state changes only on a confirmed event, and only after reconciling
// against the API — the webhook body is a hint, the API is the source of truth.
public final class WebhookProcessor {

  private static final Map<Family, String> RESOURCE_PATH =
      Map.of(
          Family.TRANSACTION, "/v2/transactions",
          Family.SUBSCRIPTION, "/v2/subscriptions",
          Family.TRANSFER, "/v2/transfers");

  private WebhookProcessor() {}

  public static void processEvent(WebhookEvent event, PagouHttpClient client) {
    if (!WebhookHandlers.isConfirmedStateChange(event.eventType())) {
      Logger.info("Ignoring non-confirming event " + event.eventType() + " (" + event.id() + ")");
      return;
    }
    if (event.resourceId() == null || event.resourceId().isEmpty()) {
      Logger.warn("Confirmed event " + event.eventType() + " without a resource id — cannot reconcile.");
      return;
    }

    try {
      JsonNode data =
          client
              .requestData(Request.get(RESOURCE_PATH.get(event.family()) + "/" + event.resourceId()), JsonNode.class)
              .data();
      String status = data != null && data.hasNonNull("status") ? data.get("status").asText() : "unknown";
      WebhookStore.setResourceState(event.resourceId(), status);
      Logger.info("Reconciled " + event.family() + " " + event.resourceId() + " → " + status);
    } catch (NotFoundException e) {
      Logger.warn("Resource " + event.resourceId() + " not found during reconciliation.");
    } catch (RuntimeException e) {
      // Reconciliation failed after the ack: a real system would requeue this
      // event for a later retry rather than replaying side effects.
      Logger.error("Reconciliation failed for " + event.id(), Map.of("message", String.valueOf(e.getMessage())));
      throw e;
    }
  }
}
